### PR TITLE
Upgrade loadflow result api

### PR DESCRIPTION
**Pull request by AAJELLAL**

### CODE REVIEW

File: src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java [hunk 1]
```diff
@@ -32,12 +37,27 @@ private static <T> Specification<T> buildSpecifications(
...
+            boolean withJoin
     ) {
         return (root, query, criteriaBuilder) -> {
             List<Predicate> predicates = new ArrayList<>();
-            filters.forEach(filter -> SpecificationUtils.addPredicate(criteriaBuilder, root, predicates, filter));
+            // user filters on main entity
+            filters.stream().filter(SpecificationBuilder::isParentFilter).forEach(filter -> SpecificationUtils.addPredicate(criteriaBuilder, root, predicates, filter));
             predicates.add(root.get(fieldName).get(uuidFieldName).in(List.of(uuid)));
+            List<ResourceFilter> childrenFilters = filters.stream().filter(resourceFilter -> !isParentFilter(resourceFilter)).toList();
+            if (withJoin) {
```
Comment (anistouri) [inline src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java:49]: I see that withJoin is used for LoadFlowResult. Why not create a separate buildSpecification, one for LimitViolation and one for LoadFlowResult, and factorize the common code?

File: src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java [hunk 2]
```diff
@@ -32,12 +37,27 @@ private static <T> Specification<T> buildSpecifications(
...
+                if (!childrenFilters.isEmpty()) {
+                    // user filters on OneToMany collection - needed here to filter main entities that would have empty collection when filters are applied
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java]: Fixed.

File: src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java (hunk 1 above)
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java:49]: WithJoin is used for ComponentResultEntity not for LoadFlowResult. 
buildSpecification is already used for both buildLimitViolationsSpecifications and buildLoadflowResultSpecifications  so i'm not sure that i get you point.

File: src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java [hunk 3]
```diff
@@ -32,12 +37,27 @@ private static <T> Specification<T> buildSpecifications(
             UUID uuid,
             List<ResourceFilter> filters,
             String fieldName,
-            String uuidFieldName
+            String uuidFieldName,
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/utils/SpecificationBuilder.java:40]: I didn't change this.I only added withJoin.

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java [hunk 4]
```diff
@@ -134,8 +150,14 @@ public LoadFlowResult getResult(UUID resultUuid, String stringFilters, Sort sort
...
         List<ComponentResultEntity> componentResults = getResultRepository().findComponentResults(resultUuid, fromStringFiltersToDTO(stringFilters), sort);
+        boolean hasChildFilter = fromStringFiltersToDTO(stringFilters).stream().anyMatch(resourceFilter -> !SpecificationBuilder.isParentFilter(resourceFilter));
+        boolean hasFilter = !StringUtils.isEmpty(stringFilters);
+        List<SlackBusResultEntity> slackBusResultEntities = new ArrayList<>();
+        if (hasChildFilter) {
+            slackBusResultEntities.addAll(getResultRepository().findSlackBusResults(componentResults, fromStringFiltersToDTO(stringFilters), sort));
+        }
         loadFlowResultEntity.setComponentResults(componentResults);
-        loadFlowResult = fromEntity(loadFlowResultEntity);
+        loadFlowResult = fromEntity(loadFlowResultEntity, slackBusResultEntities, hasFilter && hasChildFilter);
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java]: Indeed, I used hasFilter when I started the devs but now it's useless.

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java [hunk 5]
```diff
@@ -134,8 +150,14 @@ public LoadFlowResult getResult(UUID resultUuid, String stringFilters, Sort sort
             return null;
         }
         List<ComponentResultEntity> componentResults = getResultRepository().findComponentResults(resultUuid, fromStringFiltersToDTO(stringFilters), sort);
+        boolean hasChildFilter = fromStringFiltersToDTO(stringFilters).stream().anyMatch(resourceFilter -> !SpecificationBuilder.isParentFilter(resourceFilter));
+        boolean hasFilter = !StringUtils.isEmpty(stringFilters);
+        List<SlackBusResultEntity> slackBusResultEntities = new ArrayList<>();
+        if (hasChildFilter) {
+            slackBusResultEntities.addAll(getResultRepository().findSlackBusResults(componentResults, fromStringFiltersToDTO(stringFilters), sort));
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java]: Fixed.

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java [hunk 6]
```diff
@@ -134,8 +150,14 @@ public LoadFlowResult getResult(UUID resultUuid, String stringFilters, Sort sort
             return null;
         }
         List<ComponentResultEntity> componentResults = getResultRepository().findComponentResults(resultUuid, fromStringFiltersToDTO(stringFilters), sort);
+        boolean hasChildFilter = fromStringFiltersToDTO(stringFilters).stream().anyMatch(resourceFilter -> !SpecificationBuilder.isParentFilter(resourceFilter));
+        boolean hasFilter = !StringUtils.isEmpty(stringFilters);
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java]: I didn't get your point here.

File: src/main/java/org/gridsuite/loadflow/server/repositories/LoadFlowResultRepository.java [hunk 7]
```diff
@@ -50,17 +50,15 @@ private static LoadFlowResultEntity toResultEntity(UUID resultUuid, LoadFlowResu
...
-                .status(componentResult.getStatus())
-                .iterationCount(componentResult.getIterationCount())
-                .slackBusId(componentResult.getSlackBusId())
-                .iterationCount(componentResult.getIterationCount())
-                .slackBusId(componentResult.getSlackBusId())
-                .slackBusActivePowerMismatch(componentResult.getSlackBusActivePowerMismatch())
-                .distributedActivePower(componentResult.getDistributedActivePower())
-                .loadFlowResult(LoadFlowResultEntity.builder().resultUuid(resultUuid).build())
-                .build();
+        return new ComponentResultEntity(componentResult.getConnectedComponentNum(), componentResult.getSynchronousComponentNum(),
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/repositories/LoadFlowResultRepository.java]: Done.

File: src/main/java/org/gridsuite/loadflow/server/entities/ComponentResultEntity.java [hunk 8]
```diff
@@ -24,6 +24,17 @@
 @NoArgsConstructor
 @Entity
 public class ComponentResultEntity {
+    public ComponentResultEntity(int connectedComponentNum, int synchronousComponentNum, LoadFlowResult.ComponentResult.Status status, int iterationCount, double distributedActivePower, List<SlackBusResultEntity> slackBusResults, LoadFlowResultEntity loadFlowResult) {
```
[the remark this answers is not in the document]
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/loadflow/server/entities/ComponentResultEntity.java]: I kept the same pattern as the other calculation server but I agree the builder is much better.

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (anistouri) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowService.java]: You don't need it anymore now, so it's ok

**Review by souissimai — APPROVED** [review]

**Review by souissimai — APPROVED** [review]

**Review by anistouri — APPROVED** [review]

**Pull request merged**